swisscovid-cn-sdk-backend/swisscovid-cn-sdk-backend-ws/src/main/java/ch/ubique/swisscovid/cn/sdk/backend/ws/insertmanager/insertfilters/OverlappingIntervalsFilter.java

```diff
@@ -3,38 +3,98 @@
 import java.time.LocalDateTime;
+import java.util.ArrayList;
 import java.util.List;
 
 /**
- * A person can't be in two places at the same time. UploadVenueInfo objects whose time windows
- * overlap are both removed.
+ * A person can't be in two places at the same time. The filter checks the VenueInfo objects for
+ * integrity.
+ *
+ * <ol>
+ *   <li>The VenueInfo objects are iterated in the exact order in which they arrived in the request,
+ *       concatenating their time intervals until the preid changes or the time intervals aren't
+ *       continuous
+ *   <li>30min are subtracted from the resulting interval to account for those added upon upload
+ *   <li>The interval is added to a list
+ *   <li>Finally, once all VenueInfo objects have been processed, the list is checked for
+ *       inconsistencies, i.e. overlapping intervals
+ * </ol>
  */
 public class OverlappingIntervalsFilter implements UploadInsertionFilter {
-  @Override
-  public List<UserUploadPayloadOuterClass.UploadVenueInfo> filter(
-      LocalDateTime now,
-      List<UploadVenueInfo> uploadVenueInfoList,
-      Object principal)
-      throws InsertException {
-    for(var i = 0; i < uploadVenueInfoList.size() - 1; i++) {
-      final var visit = uploadVenueInfoList.get(i);
-      for(var j = i + 1; j < uploadVenueInfoList.size(); j++) {
-        if(doOverlap(visit, uploadVenueInfoList.get(j))) {
-          throw new OverlappingIntervalsException();
+    @Override
+    public List<UserUploadPayloadOuterClass.UploadVenueInfo> filter(
+            LocalDateTime now, List<UploadVenueInfo> uploadVenueInfoList, Object principal)
+            throws InsertException {
+        if (!uploadVenueInfoList.isEmpty()) {
+            List<TimeInterval> timeIntervals = getTimeIntervalList(uploadVenueInfoList);
+            for (var i = 0; i < timeIntervals.size() - 1; i++) {
+                final var t1 = timeIntervals.get(i);
+                for (var j = i + 1; j < timeIntervals.size(); j++) {
+                    if (TimeInterval.doIntersect(t1, timeIntervals.get(j))) {
+                        throw new OverlappingIntervalsException();
+                    }
+                }
+            }
+        }
+        return uploadVenueInfoList;
+    }
+
+    private List<TimeInterval> getTimeIntervalList(List<UploadVenueInfo> uploadVenueInfoList) {
+        List<TimeInterval> timeIntervals = new ArrayList<>();
+        var timeInterval =
+                new TimeInterval(
+                        uploadVenueInfoList.get(0).getIntervalStartMs(),
+                        uploadVenueInfoList.get(0).getIntervalStartMs());
+        for (var i = 0; i < uploadVenueInfoList.size(); i++) {
+            final var venueInfo = uploadVenueInfoList.get(i);
+            timeInterval.setIntervalEndMs(venueInfo.getIntervalEndMs());
+            if (i == uploadVenueInfoList.size() - 1) {
+                timeInterval.setIntervalEndMs(timeInterval.getIntervalEndMs() - 30 * 60 * 1000L);
```

Comment:
Maybe extract this constant to be a class constant. So that it is clear that we'll need to modify this thing here?

Reply:
Yup good point, moved it to a global constant `INTERVAL_APPENDIX`